platform/sync-installer/sync-installer-impl/src/main/java/org/codice/ddf/sync/installer/impl/SynchronisedInstallerImpl.java

```diff
@@ -0,0 +1,566 @@
+import org.osgi.service.cm.ConfigurationEvent;
+import org.osgi.service.cm.ConfigurationListener;
+import org.osgi.util.tracker.ServiceTracker;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class SynchronisedInstallerImpl implements SynchronisedInstaller {
+
+  private static final Map<Integer, String> BUNDLE_STATES =
+      new ImmutableMap.Builder<Integer, String>()
+          .put(Bundle.UNINSTALLED, "UNINSTALLED")
+          .put(Bundle.INSTALLED, "INSTALLED")
+          .put(Bundle.RESOLVED, "RESOLVED")
+          .put(Bundle.STARTING, "STARTING")
+          .put(Bundle.STOPPING, "STOPPING")
+          .put(ACTIVE, "ACTIVE")
+          .build();
+
+  private static final String SERVICE_PID_FILTER = "(" + Constants.SERVICE_PID + "=%s)";
+
+  private static final long DEFAULT_MAX_FEATURE_WAIT = TimeUnit.MINUTES.toMillis(10);
+
+  private static final long DEFAULT_MAX_SERVICE_WAIT = TimeUnit.MINUTES.toMillis(10);
+
+  private static final long DEFAULT_MAX_BUNDLE_WAIT = TimeUnit.MINUTES.toMillis(10);
+
+  private static final long DEFAULT_POLLING_INTERVAL = TimeUnit.SECONDS.toMillis(3);
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SynchronisedInstallerImpl.class);
+
+  private BundleContext bundleContext;
+
+  private ConfigurationAdmin configAdmin;
+
+  private FeaturesService featuresService;
+
+  private BundleService bundleService;
+
+  public SynchronisedInstallerImpl(
+      BundleContext bundleContext,
+      ConfigurationAdmin configAdmin,
+      FeaturesService featuresService,
+      BundleService bundleService) {
+    this.bundleContext = bundleContext;
+    this.configAdmin = configAdmin;
+    this.featuresService = featuresService;
+    this.bundleService = bundleService;
+  }
+
+  @Override
+  public void waitForBootFinish() throws SynchronisedInstallerException {
+    waitForBootFinish(DEFAULT_MAX_FEATURE_WAIT);
+  }
+
+  @Override
+  public void waitForBootFinish(long maxWaitTime) throws SynchronisedInstallerException {
+    Callable<Boolean> isBootFinishedAvailable =
+        () -> {
+          LOGGER.debug("Waiting for BootFinished service to appear.");
+          return bundleContext.getServiceReference(BootFinished.class) != null;
+        };
+
+    wait(
+        isBootFinishedAvailable,
+        maxWaitTime,
+        DEFAULT_POLLING_INTERVAL,
+        String.format("BootFinished service did not appear within [%s] ms", maxWaitTime));
+    waitForBundles();
+  }
+
+  @Override
+  public Configuration createManagedFactoryService(
+      String factoryPid, Map<String, Object> properties, String bundleLocation)
+      throws SynchronisedInstallerException {
+    return createManagedFactoryService(
+        DEFAULT_MAX_SERVICE_WAIT, factoryPid, properties, bundleLocation);
+  }
+
+  @Override
+  public Configuration createManagedFactoryService(
+      long maxWaitTime, String factoryPid, Map<String, Object> properties, String bundleLocation)
```

Comment:
`properties` and `location`  are optional and should be tagged with `@Nullable` here and in the interface.